core/src/main/java/org/apache/iceberg/SchemaParser.java

```diff
@@ -197,7 +197,12 @@ private static Type typeFromJson(JsonNode json) {
 
   private static Literal<?> defaultFromJson(String defaultField, Type type, JsonNode json) {
     if (json.has(defaultField)) {
-      return Expressions.lit(SingleValueParser.fromJson(type, json.get(defaultField)));
+      Object value = SingleValueParser.fromJson(type, json.get(defaultField));
+      if (type instanceof Types.TimestampNanoType) {
+        return Expressions.nanos((long) value);
```

Comment:
should we check value not null? the json value itself can also be null, right?

Reply:
We could precondition this, we will have slightly divergent behavior though. The normal path would bail out here

https://github.com/apache/iceberg/blob/d5476cae1b4e9b560f0126b7b9117630da9a37ca/api/src/main/java/org/apache/iceberg/expressions/Literals.java#L58C6-L61C101

So as written this would throw an NPE at a different spot only for nanos

Reply:
yes, this is not an issue. the spec says
```
Non-null values for initial-default or write-default are invalid.
```

Earlier, I wasn't sure if null values are allowed for the default attributes.